src/mesido/esdl/esdl_additional_vars_mixin.py

```diff
@@ -16,8 +16,167 @@
+                            is_pipe_dn0_needed = False
+                    except KeyError:
+                        is_pipe_dn0_needed = True
+
+                    max_demand *= 1.3  # 30% added for expected worst case heat losses
+
+                    new_pcs = []
+                    found_pc_large_enough = False
+                    for pc in self.pipe_classes(connected_asset):
+                        if not found_pc_large_enough:
+                            new_pcs.append(pc)
+                        if (
+                            new_pcs[-1].maximum_discharge
+                            * parameters[f"{asset}.cp"]
+                            * parameters[f"{asset}.rho"]
+                            * (parameters[f"{asset}.T_supply"] - parameters[f"{asset}.T_return"])
+                        ) >= max_demand:
+                            found_pc_large_enough = True
+
+                    # Remove pipe DN0 from the available pipe list, if there is always flow required
+                    # TODO: Bug to be resolved. Currently the solution is infeasible when
+                    # only 1 pipe class is available, bit it should be able to working. Then remove
+                    # the need for having 2 pipes classes available in new_pcs.
+                    if (
+                        not is_pipe_dn0_needed
+                        and len(new_pcs) > 2
+                        and new_pcs[0].maximum_discharge == 0.0
+                    ):
+                        new_pcs.remove(new_pcs[0])
+
+                    self._override_pipe_classes[connected_asset] = new_pcs
+                    if not self.is_hot_pipe(self.hot_to_cold_pipe(connected_asset)):
+                        self._override_pipe_classes[self.hot_to_cold_pipe(connected_asset)] = (
+                            new_pcs
+                        )
+
+            # Here we do the same for sources as for the sources.
+            for asset, (
+                connected_asset,
+                _orientation,
+            ) in self.energy_system_topology.sources.items():
+                if asset in self.energy_system_components.get("gas_source", []):
+                    try:
+                        max_prod_g_s = min(
+                            max(self.get_timeseries(f"{asset}.maximum_gas_source").values),
+                            bounds[f"{asset}.Gas_source_mass_flow"][1],
+                        )
+                    except KeyError:
+                        max_prod_g_s = bounds[f"{asset}.Gas_source_mass_flow"][1]
+                    new_pcs = []
+                    found_pc_large_enough = False
+                    for pc in self.gas_pipe_classes(connected_asset):
+                        if not found_pc_large_enough:
+                            new_pcs.append(pc)
+                            if (
+                                new_pcs[-1].maximum_discharge
+                                * self.parameters(0)[f"{connected_asset}.rho"]
```

Comment:
again rho is in g/s already, so no divide by 1000 needed

Reply:
Thanks. Code has been updated